### PR TITLE
feat(cosmwasm): contract builder method and cw20 build

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
for https://github.com/informalsystems/cosmos.nix/issues/172

### CODE REVIEW

**Comment by dzmitry-lahoda** [comment]
`nix fmt` behaved as it behaved. so nothing from me to fmt more files.

**Comment by dzmitry-lahoda** [comment]
cc @JonathanLorimer

**Comment by JonathanLorimer** [comment]
> `nix fmt` behaved as it behaved. so nothing from me to fmt more files.

Weird, must have been some merge skew or something. I'm surprised CI didn't fail tho...

**Review by JonathanLorimer — COMMENTED** [review]
Looks great, just left a couple of comments and nitpicks. Sorry about the formatting issues! That's my bad.

**Comment by JonathanLorimer** [comment]
> cc @JonathanLorimer

You can also request me as a reviewer in the top right and I will get an email notification!

File: lib/default.nix [hunk 1]
```diff
@@ -1,4 +1,53 @@
...
+        {
+          buildPhase = "cargo build --lib --target ${target} --profile ${profile} --package ${pname}";
+          RUSTFLAGS = "-C link-arg=-s";
+          nativeBuildInputs = wasmNativeBuildInputs;
+          installPhase = ''
+            mkdir --parents $out/lib
+            # from https://github.com/CosmWasm/rust-optimizer/blob/main/Dockerfile
+            # --signext-lowering is needed to support blockchains runnning CosmWasm < 1.3. It can be removed eventually
+            wasm-opt target/${target}/release/${binaryName} -o $out/lib/${binaryName} -Os --signext-lowering
+            cosmwasm-check $out/lib/${binaryName}
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply lib/default.nix:45]: see comments in other comment

File: lib/default.nix [hunk 2]
```diff
@@ -1,4 +1,53 @@
...
+      rustPlatform.buildRustPackage (
+        {
+          buildPhase = "cargo build --lib --target ${target} --profile ${profile} --package ${pname}";
+          RUSTFLAGS = "-C link-arg=-s";
+          nativeBuildInputs = wasmNativeBuildInputs;
+          installPhase = ''
+            mkdir --parents $out/lib
+            # from https://github.com/CosmWasm/rust-optimizer/blob/main/Dockerfile
+            # --signext-lowering is needed to support blockchains runnning CosmWasm < 1.3. It can be removed eventually
+            wasm-opt target/${target}/release/${binaryName} -o $out/lib/${binaryName} -Os --signext-lowering
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply lib/default.nix]: i tried to split onto phases and sub phases and nothing worked. what would be clean design? I do not see here design accepted by nix for wasm if any, so i ensured that people can do whatever rust/cargo fun they have to be handled.

**Comment by JonathanLorimer** [comment]
@dzmitry-lahoda I got `wasm-opt` working in the buildPhase and am going to try and get `cosmwasm-check` working in the checkPhase. Do you mind if I take some time to try and get that working? I think using the correct stdenv phases is desirable.

**Comment by JonathanLorimer** [comment]
@dzmitry-lahoda can you cherry-pick this [commit](https://github.com/informalsystems/cosmos.nix/pull/182/commits/27e6f2fe1037e38a92e11e4216a8502ed26ad910) onto your branch? This one works, you can check it by running `nix build .#cw20-base -L`

**Pull request closed without merging**